### PR TITLE
[otbn] Fix wrapping for grs1 increment with bn.lid/bn.sid

### PR DESCRIPTION
**Pull request by rswarbrick**
(EDIT: Text updated to match latest version of patch)

Because `grs1` has a 14-bit signed immediate added to it, we can have a
valid address even when `grs1` is really big. For example, suppose that
`x21` is `0x1000` in

    bn.lid   x10, -2144(x21++)

Then we should write `x21 = 0x1020` (rather than the `0x20` that we were
writing).

Similarly, if `x21` is `0xffffffc0` (2^32 - 64) then

    bn.lid   x10, 64(x21++)

loads from the valid address of 0 and should increment `x21` to
`0xffffffe0`.

Getting this right implies a 27-bit adder, plus extra muxing for the
top bits of increment_out. To avoid that, we could e.g. change the
spec to only operate on the lower 16 bits of the base address (which
would then need an 11-bit adder and no muxing on most of the upper
bits: much more manageable!)

### CODE REVIEW

**Comment by GregAC** [comment]
This comes back to the discussion here: https://github.com/lowRISC/opentitan/issues/3587 where I think the original conclusion was we'd just wrap at the DMem boundary, potentially triggering an error if we see an overflow but we never concluded on that and didn't consider this case.

I can see how fault on DMem wrap when incrementing would be awkward as you could issue a `BN.LID` with some out of range base address, using the immediate to bring it into range then immediately fault if you do the same but with an increment or get weird behaviour if there's no fault as the manual add would do something you'd expect and the wrapping increment would not.

So what you suggest seems reasonable. Another option would be to say no out of range base addresses full-stop and fault on that. @felixmiller @imphil any thoughts? The more conservative fault on any out of range address may help catch more SW bugs, actually using an out of range base address probably isn't something SW will rely on.

It is a bit of a shame to loose parameterization on DMem size. If we go with this we definitely want an assert that will fail if DMem is too large for the RTL.

**Comment by rswarbrick** [comment]
Oh! I didn't realise that the question was still open! (I think I'd got confused by the fact we decided what to do for loop handling). I think I like the instruction semantics from this patch (because the documentation won't need to say "size of Dmem" anywhere when describing the increment behaviour), but you're right: we should track this question in #3587.

I'll convert this PR to a draft, since we shouldn't merge it before we've decided what we're doing. Let's try to get consensus on #3587 soonish (or I'll alter the ISS to match the RTL, but it always seems a bit naff to do things "backwards" like that).

**Comment by imphil** [comment]
Looping back the discussion in #3587 to this PR:
* The implementation here matches the documentation we already have.
* Let's await the final discussion in the OTBN sprint meeting today, but I'm happy with the approach this PR is taking.
* Could you please add an assert and a small comment on what to do if the DMEM size gets increased beyond 16 kiB?

**Comment by rswarbrick** [comment]
Gaargh. This isn't quite as easy as I thought. The patch works fine for large base addresses with negative offsets, but it doesn't deal properly with things like a base address of `0xffffffff` and an offset of `33`. As far as I can see, dealing with that sort of case properly really needs a full 32-bit adder (or at least a 32+5 bit adder or something).

I'm going to push up a version with that for now, just to get everything back in sync. Let's loop back to the spec question (maybe in #3587, or maybe in a new issue). We've already got open questions about the bottom of the WDR index. Maybe we should decide to do something like just operate on the bottom 16 bits of the base address too.

**Review by GregAC — APPROVED** [review]

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (imphil) [inline hw/ip/otbn/rtl/otbn_controller.sv:339]: That'll be a width mismatch in lint. Make it `'h1`, or specify the width explicitly to match the other operand (26'h1)

**Review by imphil — APPROVED** [review]

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (rswarbrick) [reply hw/ip/otbn/rtl/otbn_controller.sv:339]: Do we trigger lint warnings for the previous two statements? I followed their style.

**Pull request merged**
